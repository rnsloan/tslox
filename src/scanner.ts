export enum TokenType {
  // one character tokens
  LEFT_PAREN = "LEFT_PAREN",
  RIGHT_PAREN = "RIGHT_PAREN",
  LEFT_BRACE = "LEFT_BRACE",
  RIGHT_BRACE = "RIGHT_BRACE",
  COMMA = "COMMA",
  DOT = "DOT",
  MINUS = "MINUS",
  PLUS = "PLUS",
  SEMICOLON = "SEMICOLON",
  SLASH = "SLASH",
  STAR = "STAR",

  // One or two character tokens
  BANG = "BANG",
  BANG_EQUAL = "BANG_EQUAL",
  EQUAL = "EQUAL",
  EQUAL_EQUAL = "EQUAL_EQUAL",
  GREATER = "GREATER",
  GREATER_EQUAL = "GREATER_EQUAL",
  LESS = "LESS",
  LESS_EQUAL = "LESS_EQUAL",

  // Literals
  IDENTIFIER = "IDENTIFIER",
  STRING = "STRING",
  NUMBER = "NUMBER",

  // Keywords
  AND = "AND",
  CLASS = "CLASS",
  ELSE = "ELSE",
  FALSE = "FALSE",
  FUN = "FUN",
  FOR = "FOR",
  IF = "IF",
  NIL = "NIL",
  OR = "OR",
  PRINT = "PRINT",
  RETURN = "RETURN",
  SUPER = "SUPER",
  THIS = "THIS",
  TRUE = "TRUE",
  VAR = "VAR",
  WHILE = "WHILE",

  EOF = "EOF",
}

const IDENTIFIERS: { [key: string]: TokenType } = {
  "and": TokenType.AND,
  "class": TokenType.CLASS,
  "else": TokenType.ELSE,
  "false": TokenType.FALSE,
  "fun": TokenType.FUN,
  "for": TokenType.FOR,
  "if": TokenType.IF,
  "nil": TokenType.NIL,
  "or": TokenType.OR,
  "print": TokenType.PRINT,
  "return": TokenType.RETURN,
  "super": TokenType.SUPER,
  "this": TokenType.THIS,
  "true": TokenType.TRUE,
  "var": TokenType.VAR,
  "while": TokenType.WHILE,
};

export interface IToken {
  type: TokenType;
  lexeme: string | number;
  literal: string | number | boolean | null;
  line: number;
  start: number;
  end: number;
}

interface AddToken extends Partial<IToken> {
  type: TokenType;
}

const isDigit = (c: string) => c?.match(/[0-9]/);
const isAlpha = (c: string) => c?.match(/_|[a-z]/i);
const isAlphaNumeric = (c: string) => isDigit(c) || isAlpha(c);

export function scanner(input: string): IToken[] {
  const output: IToken[] = [];
  const sep = input.split("");
  let position = 0;
  let line = 1;

  function addToken({
    type,
    lexeme = sep[position],
    literal = null,
    start = position,
    end = position,
  }: AddToken) {
    output.push({ type, line, lexeme, literal, start, end });
  }

  function match(character: string): boolean {
    if (sep[position + 1] === character) {
      position++;
      return true;
    }
    return false;
  }

  function peek(): string {
    return sep[position + 1];
  }

  function advance() {
    position++;
  }

  function string(quoteCharacter: string) {
    const start = position;

    advance();

    while (sep[position] !== quoteCharacter) {
      advance();
    }

    const end = position;

    addToken({
      type: TokenType.STRING,
      lexeme: input.slice(start + 1, end),
      start: start + 1,
      end: end - 1,
    });
  }

  function number() {
    const start = position;

    while (isDigit(peek()) || peek() === ".") {
      advance();
    }

    const end = position;

    addToken({
      type: TokenType.NUMBER,
      lexeme: input.slice(start, end + 1),
      literal: Number(input.slice(start, end + 1)),
      start,
      end,
    });
  }

  function identifier() {
    const start = position;

    advance();

    while (isAlphaNumeric(peek())) {
      advance();
    }

    const end = position;
    const value = input.slice(start, end + 1);
    let tokenType = TokenType.IDENTIFIER;
    let literal;

    if (IDENTIFIERS[value]) {
      tokenType = IDENTIFIERS[value];
    }

    if (tokenType === TokenType.TRUE) {
      literal = true;
    }

    if (tokenType === TokenType.FALSE) {
      literal = false;
    }

    addToken({
      type: tokenType,
      lexeme: value,
      literal,
      start: start,
      end: end,
    });
  }

  while (position < sep.length) {
    switch (sep[position]) {
      // one character tokens
      case "(":
        addToken({ type: TokenType.LEFT_PAREN });
        break;
      case ")":
        addToken({ type: TokenType.RIGHT_PAREN });
        break;
      case "{":
        addToken({ type: TokenType.LEFT_BRACE });
        break;
      case "}":
        addToken({ type: TokenType.RIGHT_BRACE });
        break;
      case ",":
        addToken({ type: TokenType.COMMA });
        break;
      case ".":
        addToken({ type: TokenType.DOT });
        break;
      case "-":
        addToken({ type: TokenType.MINUS });
        break;
      case "+":
        addToken({ type: TokenType.PLUS });
        break;
      case ";":
        addToken({ type: TokenType.SEMICOLON });
        break;
      case "*":
        addToken({ type: TokenType.STAR });
        break;
      // One or two character tokens
      case "!":
        match("=")
          ? addToken({
            type: TokenType.BANG_EQUAL,
            lexeme: `!=`,
            end: position + 1,
          })
          : addToken({ type: TokenType.BANG });
        break;
      case "=":
        match("=")
          ? addToken({
            type: TokenType.EQUAL_EQUAL,
            lexeme: `==`,
            end: position + 1,
          })
          : addToken({ type: TokenType.EQUAL });
        break;
      case "<":
        match("=")
          ? addToken({
            type: TokenType.LESS_EQUAL,
            lexeme: `<=`,
            end: position + 1,
          })
          : addToken({ type: TokenType.LESS });
        break;
      case ">":
        match("=")
          ? addToken({
            type: TokenType.GREATER_EQUAL,
            lexeme: `>=`,
            end: position + 1,
          })
          : addToken({ type: TokenType.GREATER });
        break;
      // Slash or comment
      case "/":
        if (match("/")) {
          while (peek() != "\n" && position < sep.length) {
            advance();
          }
          break;
        }
        addToken({ type: TokenType.SLASH });
        break;
      case '"':
        string('"');
        break;
      case "'":
        string("'");
        break;
      // ignored tokens
      case " ":
      case "\r":
      case "\t":
        break;
      case "\n":
        line++;
        break;
      default:
        if (isDigit(sep[position])) {
          number();
        } else if (isAlpha(sep[position])) {
          identifier();
        } else {
          throw new Error(
            `Unexpected character '${sep[position]}' {${line}:${position}}`,
          );
        }
    }

    position++;
  }

  addToken({
    type: TokenType.EOF,
    lexeme: "",
    literal: null,
    start: position + 1,
    end: position + 1,
  });
  return output;
}
