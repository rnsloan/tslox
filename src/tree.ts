// deno-lint-ignore no-explicit-any
type Data = any;

enum STRATEGY {
  DEPTH,
  BREADTH,
}

export interface INode {
  parent: Node | null;
  children: Node[];
  data: Data;
  isRoot: () => boolean;
  isLeaf: () => boolean;
  hasChildren: () => boolean;
  drop: () => void;
  addChild: (data: Data) => Node;
  addChildAtIndex: (data: Data, index: number) => Node;
  getPath: (node: Node) => Node[];
}

class Node implements INode {
  protected id: string;
  readonly parent: Node | null;
  readonly children: Node[];
  readonly data: Data;

  constructor(data: Data, parent?: Node) {
    this.id = crypto.randomUUID();
    this.parent = parent || null;
    this.data = data;
    this.children = [];
  }

  isRoot() {
    return false;
  }

  isLeaf() {
    return Boolean(!this.children.length);
  }

  hasChildren() {
    return Boolean(this.children.length);
  }

  drop(): Node | undefined {
    if (this.isRoot()) {
      console.warn("Cannot drop the Root Node");
      return;
    }

    const index = this.parent?.children.findIndex((n) => n.id === this.id);

    if (index) {
      const result = this.parent?.children.splice(index, 1);
      if (result?.length) {
        return result[0];
      }
    }

    console.error("Unable to drop Node");
  }

  addChild(data: Data) {
    const node = new Node(data, this);
    this.children.push(node);
    return node;
  }

  addChildAtIndex(data: Data, index: number) {
    const node = new Node(data, this);
    this.children.splice(index, 0, node);
    return node;
  }

  getPath() {
    interface AddParent {
      (n: Node): AddParent;
    }

    const path = [];
    path.push(this);

    const addParent = (n: Node): AddParent | undefined => {
      const { parent } = n;

      if (parent) {
        path.push(parent);
        return addParent(parent);
      }
    };

    addParent(this);
    return path.reverse();
  }
}

export interface IRootNode extends INode {
  find: (
    predicate: (n: Node) => boolean,
    strategy?: STRATEGY,
  ) => Node | undefined;
  findAll: (predicate: (n: Node) => boolean, strategy?: STRATEGY) => Node[];
}

export class RootNode extends Node implements IRootNode {
  constructor(data: Data) {
    super(data);
    this.id = `root-${crypto.randomUUID()}`;
  }

  isRoot() {
    return true;
  }

  isLeaf() {
    return false;
  }

  find(predicate: (n: Node) => boolean, strategy?: STRATEGY): Node | undefined {
    let result: Node | undefined;

    const findDfs = (node: Node) => {
      if (predicate(node) && !result) {
        result = node;
        return;
      }

      for (let i = 0; i < node.children.length; i++) {
        findDfs(node.children[i]);
      }
    };
    const findBfs = (node: Node) => {
      const queue = [node];

      while (queue.length && !result) {
        const n = queue.shift();

        if (n && predicate(n)) {
          result = n;
        }

        n?.children?.forEach((item) => queue.push(item));
      }
    };

    if (strategy === STRATEGY.BREADTH) {
      findBfs(this);
    } else {
      findDfs(this);
    }

    return result;
  }

  findAll(predicate: (n: Node) => boolean, strategy?: STRATEGY): Node[] {
    const result: Node[] = [];
    const findAllDfs = (node: Node) => {
      if (predicate(node)) {
        result.push(node);
      }

      node.children.forEach((n) => {
        findAllDfs(n);
      });
    };

    const findAllBfs = (node: Node) => {
      const queue = [node];

      while (queue.length) {
        const n = queue.shift();

        if (n && predicate(n)) {
          result.push(n);
        }

        n?.children?.forEach((item) => queue.push(item));
      }
    };

    if (strategy === STRATEGY.BREADTH) {
      findAllBfs(this);
    } else {
      findAllDfs(this);
    }

    return result;
  }
}

export default class Tree {
  static STRATEGY = STRATEGY;

  root: RootNode | null;

  constructor() {
    this.root = null;
  }

  parse(data: Data) {
    this.root = new RootNode(data);
    return this.root;
  }
}
