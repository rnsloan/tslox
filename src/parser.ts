import { IToken, TokenType } from "./scanner.ts";
import { Tree } from "./tree.ts";

export enum ASTNodeType {
  Program = "Program",
  Literal = "Literal",
  VariableDeclaration = "VariableDeclaration",
  VariableDeclarator = "VariableDeclarator",
  UnaryExpression = "UnaryExpression",
  CallExpression = "CallExpression",
  MemberExpression = "MemberExpression",
  Identifier = "Identifier",
}
export interface IProgram {
  type: ASTNodeType.Program;
  body: ASTNode[];
}
interface ILiteral {
  type: ASTNodeType.Literal;
  value: null | number | string | boolean;
  raw: null | number | string;
}
interface IIdentifier {
  type: ASTNodeType.Identifier;
  name: string;
}
interface IUnaryExpression {
  type: ASTNodeType.UnaryExpression;
  operator: string;
  prefix: boolean;
  argument: ASTNode;
}
interface ICallExpression {
  type: ASTNodeType.CallExpression;
  callee: IIdentifier;
  arguments?: ASTNode[];
}
interface IMemberExpression {
  type: ASTNodeType.MemberExpression;
  object: ILiteral;
  property: IIdentifier;
}
interface IDeclarator {
  type: ASTNodeType.VariableDeclarator;
  id: {
    type: ASTNodeType;
    name: string | number;
  };
  init: ASTNode;
}
interface IVariableDeclaration {
  type: ASTNodeType.VariableDeclaration;
  declarations: IDeclarator[];
}

export type ASTNode =
  | IProgram
  | IDeclarator
  | IVariableDeclaration
  | IUnaryExpression
  | ICallExpression
  | IMemberExpression
  | ILiteral
  | IIdentifier;

let code: IToken[];
let position = 0;

function match(token: IToken, comparison: TokenType | TokenType[]): boolean {
  return Array.isArray(comparison)
    ? comparison.includes(token.type)
    : token.type === comparison;
}

function advance(amount = 1): IToken {
  position = position + amount;
  return code[position];
}

function peek(amount = 1): IToken {
  return code[position + amount];
}

function evalStatement(): ASTNode | null {
  return evalExpresion();
}

function evalExpresion(): ASTNode | null {
  return evalAssignment();
}

function evalAssignment(): ASTNode | null {
  // TODO: Implement ( call "." )? IDENTIFIER "=" assignment
  //const primary = evalPrimary();
  return evalLogicOr();
}

// TODO complete implementations
function evalLogicOr(): ASTNode | null {
  return evalEquality();
}

function evalEquality(): ASTNode | null {
  return evalComparison();
}

function evalComparison(): ASTNode | null {
  return evalTerm();
}

function evalTerm(): ASTNode | null {
  return evalFactor();
}

function evalFactor(): ASTNode | null {
  // TODO implement factor
  // need to handle expression "(" expression ")"
  return evalUnary();
}

function evalUnary(): ASTNode | null {
  if (
    match(code[position], [
      TokenType.MINUS,
      TokenType.BANG,
    ])
  ) {
    const unaryToken = code[position];

    advance();

    const unary: IUnaryExpression = {
      type: ASTNodeType.UnaryExpression,
      operator: match(unaryToken, [
          TokenType.MINUS,
        ])
        ? "-"
        : "!",
      prefix: true,
      argument: evalUnary() as ASTNode,
    };

    return unary;
  }
  return evalCall();
}

function evalCall(): ILiteral | IIdentifier | ICallExpression | null {
  // TODO: primary "." IDENTIFIER
  const primary = evalPrimary();
  if (primary === null) {
    return null;
  }

  if (
    primary.type === ASTNodeType.Identifier &&
    match(code[position], TokenType.LEFT_PAREN)
  ) {
    const callExpression: ICallExpression = {
      type: ASTNodeType.CallExpression,
      callee: primary,
    };
    const args: ASTNode[] = [];

    while (code[position] && !match(code[position], TokenType.RIGHT_PAREN)) {
      if (match(code[position], [TokenType.LEFT_PAREN, TokenType.COMMA])) {
        advance();
        continue;
      }

      const primary = evalPrimary();

      if (primary) {
        args.push(primary);
      } else {
        advance();
      }
    }

    if (args.length) {
      callExpression.arguments = args;
    }

    // skip ")"
    advance();

    return callExpression;
  }

  return primary;
}

function evalPrimary(): ILiteral | IIdentifier | null {
  const token = code[position];

  if (
    match(token, [
      TokenType.TRUE,
      TokenType.FALSE,
      TokenType.THIS,
      TokenType.NUMBER,
      TokenType.STRING,
      TokenType.IDENTIFIER,
      TokenType.SUPER,
      TokenType.NIL,
    ])
  ) {
    advance();

    if (match(token, TokenType.NIL)) {
      return {
        type: ASTNodeType.Literal,
        value: null,
        raw: null,
      };
    }

    if (match(token, TokenType.IDENTIFIER)) {
      return {
        type: ASTNodeType.Identifier,
        name: token.lexeme as string,
      };
    }

    return {
      type: ASTNodeType.Literal,
      value: token.literal as boolean | string | number,
      raw: `${token.lexeme}`,
    };
  }

  /*
  if (match(token, TokenType.LEFT_PAREN)) {
    const codeSegment: IToken[] = [];
    let i = position + 1;
    let leftParensCount = 0;

    while (!match(code[i], TokenType.RIGHT_PAREN) && leftParensCount === 0) {
      if (match(code[i], TokenType.EOF)) {
        throw new Error("Expected Right Parenthesis not found");
      }

      if (match(code[i], TokenType.LEFT_PAREN)) {
        leftParensCount++;
      }

      if (match(code[i], TokenType.RIGHT_PAREN)) {
        leftParensCount--;
      }

      codeSegment.push(code[i]);
      i++;
    }

    // TODO - fix checkExpresion() to work with code segment
    if (evalExpresion()) {
      advance(i);

      return {
        type: ASTNodeType.Literal,
        value: token.literal as boolean | string | number,
        raw: `${token.lexeme}`,
      };
    }
  }
  */
  return null;
}

export function parser(c: IToken[]) {
  code = c;
  position = 0;
  const tree = new Tree();
  const root = tree.parse({ type: "Program" });

  /*
  function addSibling(data: IASTNode) {
    if (node) {
      node.parent?.addChild(data);
    }
    root.addChild(data);
  }

  function addChild(data: IASTNode) {
    node = node ? node.addChild(data) : root.addChild(data);
  }
  */

  function walk(): ASTNode | null {
    let token = code[position];

    if (match(token, TokenType.SEMICOLON)) {
      advance();
      return null;
    }

    if (match(token, TokenType.VAR)) {
      const astNode: IVariableDeclaration = {
        type: ASTNodeType.VariableDeclaration,
        declarations: [],
      };

      token = advance();

      if (!match(token, TokenType.IDENTIFIER)) {
        throw new Error(
          `Identifier expected got ${token.type} {${token.line}:${token.start}}`,
        );
      }

      const declarator: IDeclarator = {
        type: ASTNodeType.VariableDeclarator,
        id: {
          type: ASTNodeType.Identifier,
          name: token.lexeme,
        },
        init: {
          type: ASTNodeType.Literal,
          value: null,
          raw: null,
        },
      };

      if (match(peek(), TokenType.EQUAL)) {
        token = advance(2);
        const val = evalExpresion();
        if (val) {
          declarator.init = val;
        } else {
          throw new Error(
            `Unable to parse variable value {${token.line}:${token.start}}`,
          );
        }
      }

      astNode.declarations.push(declarator);

      return astNode;
    }

    return evalStatement();
  }

  // walk() function advances position value
  while (position < c.length - 1) {
    const astNode = walk();

    if (astNode) {
      root.addChild(astNode);
    }
  }

  return tree;
}
