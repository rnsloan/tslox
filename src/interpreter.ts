import { INode, Tree } from "./tree.ts";
import { ASTNodeType } from "./parser.ts";

function codeGenerate(node: INode): string {
  let code = "";
  const { data } = node;

  if (!data || !data.type) {
    console.error(node);
    throw new Error("ASTNode missing data or data.type property");
  }

  const { type } = data;

  switch (type) {
    case ASTNodeType.Program: {
      code = node.children.map((n) => codeGenerate(n)).join("\n");
      break;
    }
    case ASTNodeType.VariableDeclaration: {
      code = `var ${data.declarations[0].id.name} = ${
        data.declarations[0].init.value
      };`;
      break;
    }
    default: {
      throw new Error(`Unrecognised AST type ${type}`);
    }
  }

  return code;
}

export function interpreter(ast: Tree): string {
  return codeGenerate(ast.root as INode);
}
