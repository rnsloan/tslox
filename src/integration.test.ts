import { assertSnapshot } from "https://deno.land/std@0.143.0/testing/snapshot.ts";
import { scanner } from "./scanner.ts";
import { ASTNode, parser } from "./parser.ts";
import { interpreter } from "./interpreter.ts";
import { Tree } from "./tree.ts";

function test(code: string) {
  const tokens = scanner(code);
  const ast = parser(tokens);
  return interpreter(ast as Tree<ASTNode>);
}

Deno.test("Integration", async (t) => {
  const code = `
var foo = bar;
var baz = 'hello world';
var alpha = 5454.454;
var beta = null;
var gamma = nil;
`;
  await assertSnapshot(t, test(code));

  const code2 = `
foo(true);
bar();
var foo = hello(34,"world");
`;
  await assertSnapshot(t, test(code2));

  const code3 = `
var foo = -5;
var bar = !false
var baz = !!true
`;
  await assertSnapshot(t, test(code3));

  const code4 = `
var foo = 5 * 10
var bar = 5*10/2;
var baz = 5*(10/2);
`;
  await assertSnapshot(t, test(code4));

  const code5 = `
var foo = 2+10/2
var bar = (2+10)/2
`;
  await assertSnapshot(t, test(code5));

  const code6 = `
10 != 6
'a' >= 'b'
6 < 4
foo != bar
`;
  await assertSnapshot(t, test(code6));

  const code7 = `
6 or null;
(6 and null) or bar;
  `;
  await assertSnapshot(t, test(code7));

  const code8 = `
var one = foo.bar;
var two = foo.bar.baz.troll;
  `;
  await assertSnapshot(t, test(code8));
});
