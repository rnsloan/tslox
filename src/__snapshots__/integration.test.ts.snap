export const snapshot = {};

snapshot[`Integration 1`] = `
'var foo = bar;
var baz = "hello world";
var alpha = 5454.454;
var beta = null;
var gamma = null;'
`;

snapshot[`Integration 2`] = `
'foo(true)
bar()
var foo = hello(34,"world");'
`;

snapshot[`Integration 3`] = `
"var foo = -5;
var bar = !false;
var baz = !!true;"
`;

snapshot[`Integration 4`] = `
"var foo = 5 * 10;
var bar = 5 * (10 / 2);
var baz = 5 * (10 / 2);"
`;

snapshot[`Integration 5`] = `
"var foo = 2 + (10 / 2);
var bar = (2 + 10) / 2;"
`;

snapshot[`Integration 6`] = `
'10 != 6
"a" >= "b"
6 < 4
foo != bar'
`;

snapshot[`Integration 7`] = `
"6 || null
(6 && null) || bar"
`;

snapshot[`Integration 8`] = `
"var one = foo.bar;
var two = foo.bar.baz.troll;"
`;

snapshot[`Integration 9`] = `
"var one = newPoint(2,0).y;
var two = foo.bar().baz;
var three = foo().bar().baz();
var four = foo.bar.baz();"
`;

snapshot[`Integration 10`] = `"newPoint(x,2,0).y = 3"`;

snapshot[`Integration 11`] = `
"{
  newPoint(x,2,0).y = 3;var tan = 7;
}"
`;

snapshot[`Integration 12`] = `
"while (foo < 4) {
  i + 1;foo + 1
}"
`;

snapshot[`Integration 13`] = `"while (i < 10) i + 1"`;

snapshot[`Integration 14`] = `"return foo < bar;"`;

snapshot[`Integration 15`] = `'console.log("hello world")'`;
